packages/multichain/src/caip25Permission.ts

```diff
@@ -0,0 +1,269 @@
+  PermissionSpecificationBuilder,
+  EndowmentGetterParams,
+  ValidPermissionSpecification,
+  PermissionValidatorConstraint,
+  PermissionConstraint,
+} from '@metamask/permission-controller';
+import {
+  CaveatMutatorOperation,
+  PermissionType,
+} from '@metamask/permission-controller';
+import type { CaipAccountId, Json } from '@metamask/utils';
+import {
+  parseCaipAccountId,
+  type Hex,
+  type NonEmptyArray,
+} from '@metamask/utils';
+import { strict as assert } from 'assert';
+import { cloneDeep, isEqual } from 'lodash';
+
+import { assertScopesSupported } from './scope/assert';
+import { validateAndFlattenScopes } from './scope/authorization';
+import type {
+  ExternalScopeString,
+  ScopeObject,
+  ScopesObject,
+} from './scope/types';
+
+export type Caip25CaveatValue = {
+  requiredScopes: ScopesObject;
+  optionalScopes: ScopesObject;
+  sessionProperties?: Record<string, Json>;
+  isMultichainOrigin: boolean;
+};
+
+export const Caip25CaveatType = 'authorizedScopes';
+
+export const Caip25CaveatFactoryFn = (value: Caip25CaveatValue) => {
+  return {
+    type: Caip25CaveatType,
+    value,
+  };
+};
+
+export const Caip25EndowmentPermissionName = 'endowment:caip25';
+
+type Caip25EndowmentSpecification = ValidPermissionSpecification<{
+  permissionType: PermissionType.Endowment;
+  targetName: typeof Caip25EndowmentPermissionName;
+  endowmentGetter: (_options?: EndowmentGetterParams) => null;
+  validator: PermissionValidatorConstraint;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+/**
+ * `endowment:caip25` returns nothing atm;
+ *
+ * @param builderOptions - The specification builder options.
+ * @param builderOptions.findNetworkClientIdByChainId - The hook to find the networkClientId for a chainId.
+ * @returns The specification for the `caip25` endowment.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.Endowment,
+  // TODO: FIX THIS
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  any,
+  Caip25EndowmentSpecification
+> = ({
+  findNetworkClientIdByChainId,
+}: {
+  findNetworkClientIdByChainId: (chainId: Hex) => NetworkClientId;
+}) => {
+  return {
+    permissionType: PermissionType.Endowment,
+    targetName: Caip25EndowmentPermissionName,
+    allowedCaveats: [Caip25CaveatType],
+    endowmentGetter: (_getterOptions?: EndowmentGetterParams) => null,
+    validator: (permission: PermissionConstraint) => {
+      const caip25Caveat = permission.caveats?.[0];
+      if (
+        permission.caveats?.length !== 1 ||
+        caip25Caveat?.type !== Caip25CaveatType
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Invalid caveats. There must be a single caveat of type "${Caip25CaveatType}".`,
+        );
+      }
+
+      // TODO: FIX THIS TYPE
+      const { requiredScopes, optionalScopes, isMultichainOrigin } = (
+        caip25Caveat as unknown as { value: Caip25CaveatValue }
+      ).value;
```

Comment:
I don't think we still have to cast this to unknown first?
```suggestion
      const { requiredScopes, optionalScopes, isMultichainOrigin } =
        caip25Caveat.value as Caip25CaveatValue;
```

Reply:
fixed https://github.com/MetaMask/core/pull/4784/commits/597e837720641ec9d567622e2a61343aaa762b55